src/main/java/org/wildfly/security/keystore/KeyStoreUtil.java

```diff
@@ -0,0 +1,166 @@
+ *
+ * @author <a href="mailto:mmazanek@redhat.com">Martin Mazanek</a>
+ */
+public class KeyStoreUtil {
+
+    private static final String BCFKS = "BCFKS";
+    private static final String BKS = "BKS";
+    private static final String JCEKS = "JCEKS";
+    private static final String JKS = "JKS";
+    private static final String PKCS12 = "PKCS12";
+    private static final String UBER = "UBER";
+
+    private static final int VERSION_0 = 0;
+    private static final int VERSION_1 = 1;
+    private static final int VERSION_2 = 2;
+    private static final int JCEKS_MAGIC = 0xcececece;
+    private static final int JKS_MAGIC = 0xfeedfeed;
+    private static final int SEQUENCE = 0x30000000;
+
+    /**
+     * Tries to parse a keystore based on known recognizable patterns.
+     *
+     * This method can parse JKS, JCEKS, PKCS12, BKS, BCFKS and UBER key stores.
+     * At first the method looks for recognizable patterns of JKS, JCEKS, PKCS12 and BKS key store types and tries to parse them if found.
+     * If the patter recognition failed, brute force is used to load the key store.
+     *
+     * The provider supplier is used for loading the key stores.
+     *
+     * @param providers provider supplier for loading the keystore (must not be {@code null})
+     * @param providerName if specified only providers with this name will be used
+     * @param is the key store file input stream (must not be {@code null})
+     * @param filename the filename for prioritizing brute force checks using the file extension
+     * @param password password of the key store
+     * @return loaded key store if recognized
+     * @throws IOException
+     */
+    public static KeyStore loadKeyStore(final Supplier<Provider[]> providers, final String providerName, FileInputStream is, String filename, char[] password) throws IOException {
+
+        DataInputStream dis = new ResetableDataFileInputStream(is);
+
+        int firstInt = dis.readInt();
+        dis.reset();
+
+        KeyStore result = null;
+
+        try {
+            if (firstInt == JKS_MAGIC) {
+                result = tryLoadKeystore(providers, providerName, dis, password, JKS);
+            } else if (firstInt == JCEKS_MAGIC) {
+                result = tryLoadKeystore(providers, providerName, dis, password, JCEKS);
+            } else if (firstInt == VERSION_1 || firstInt == VERSION_2) {
+                dis.reset();
+                dis.skip(32);
+                byte firstElementType = dis.readByte();
+                dis.reset();
+                if (firstElementType <= 5) {
+                    result = tryLoadKeystore(providers, providerName, dis, password, BKS, UBER);
+                } else {
+                    result = tryLoadKeystore(providers, providerName, dis, password, UBER, BKS);
+                }
+
+            } else if (firstInt == VERSION_0) {
+                result = tryLoadKeystore(providers, providerName, dis, password, UBER);
```

Comment:
Based on https://github.com/wildfly/wildfly-proposals/pull/124/commits/a313b5bd33b4aee706416550c1d1227ffbc69ec2#diff-02287f425f424b348e7faba51ae7acebR80 UBER cant be determined based on pattern. But have to be tried.